spec/lib/payment_processor/go_cardless/transaction_spec.rb

```diff
@@ -0,0 +1,103 @@
+require 'rails_helper'
+require_relative 'transaction_and_subscription_examples'
+
+module PaymentProcessor
+  module GoCardless
+    describe Transaction do
+      describe '.make_transaction' do
+
+        before do
+          allow(Payment::GoCardless).to receive(:write_transaction).and_return(local_transaction)
+          allow(Payment::GoCardless).to receive(:write_customer).and_return(local_customer)
+          allow(Payment::GoCardless).to receive(:write_mandate).and_return(local_mandate)
+          allow(PaymentProcessor::Currency).to receive(:convert).and_return(double(cents: amount_in_euros*100))
+
+          allow_any_instance_of(GoCardlessPro::Services::RedirectFlowsService).to receive(:complete).and_return(completed_flow)
+          allow_any_instance_of(GoCardlessPro::Services::RedirectFlowsService).to receive(:get).and_return(completed_flow)
+          allow_any_instance_of(GoCardlessPro::Services::MandatesService).to receive(:get).and_return(mandate)
+          allow_any_instance_of(GoCardlessPro::Services::PaymentsService).to receive(:create).and_return(payment)
```

Comment:
This is great, but I'm thinking about how we should test classes like this going forward. That the class touches so many different parts of the codebase makes it very hard to unit test, and all that mocking is going to be brittle. My inclination would be to rely on integration testing alone. If there's specific behaviour here which could be moved into a pure functions then I would do that and write a unit tests for them.